cranelift/codegen/src/isa/x64/inst.isle

```diff
@@ -2765,141 +2765,213 @@
 (rule (x64_xmm_load_const ty const)
       (x64_load ty (const_to_synthetic_amode const) (ExtKind.None)))
 
+
+;;;; Flag Helpers ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;
+;; These helpers are used to emit instructions that produce or consume flags.
+;; The operations used here are by no means the only ones possible; they are
+;; simply the ones currently used in Cranelift's lowerings.
+
+;; Some operations produce flags.
+(type ProduceFlagsOp (enum (Add) (Sub)))
+
+(decl x64_produce_flags (ProduceFlagsOp Type Gpr GprMemImm) ProducesFlags)
+(rule (x64_produce_flags (ProduceFlagsOp.Add) ty src1 src2)
+      (x64_add_with_flags_paired ty src1 src2))
+(rule (x64_produce_flags (ProduceFlagsOp.Sub) ty src1 src2)
+      (x64_sub_with_flags_paired ty src1 src2))
+
+(decl asm_produce_flags (AssemblerOutputs) ProducesFlags)
```

Comment:
Can we add a comment here that this should only be provided an instruction that actually produces flags, and is meant to be used by helpers that bind it with explicit instructions? In other words, it's semantically "unsafe" in the Rust sense and has to be used in the right way to uphold invariants.